tripal_chado/src/Plugin/Field/FieldType/ChadoContactByRoleTypeDefault.php

```diff
@@ -0,0 +1,303 @@
+      'action' => 'store_link',
+      'drupal_store' => TRUE,
+      'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $linker_left_col,
+      'table_alias_mapping' => [$table_alias => $linker_table],
+    ]);
+
+    // Define the link between the linker table and the object table.
+    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, $linker_fkey_column, $terms['linker_right'], [
+      'action' => 'store',
+      'drupal_store' => TRUE,
+      'path' => $table_alias . '.' . $linker_fkey_column,
+      'table_alias_mapping' => [$table_alias => $linker_table],
+      'delete_if_empty' => TRUE,
+      'empty_value' => 0,
+    ]);
+
+    // Linker type_id
+    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_type_id', $terms['linker_type_id'], [
+      'action' => 'store',
+      'drupal_store' => FALSE,
+      'path' => $table_alias . '.type_id',
+      'table_alias_mapping' => [$table_alias => $linker_table],
+      'as' => 'linker_type_id',
+    ]);
+
+    // Linker Rank
+    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_rank', $terms['linker_rank'], [
+      'action' => 'store',
+      'drupal_store' => FALSE,
+      'path' => $table_alias . '.rank',
+      'table_alias_mapping' => [$table_alias => $linker_table],
+      'as' => 'linker_rank',
+    ]);
+
+    // The object table, the destination table of the linker table
+    // The contact name
+    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'contact_name', $terms['name'], $max_lengths['name'], [
+      'action' => 'read_value',
+      'drupal_store' => FALSE,
+      'path' => $table_alias . '.' . $linker_fkey_column . '>' . $object_table . '.' . $object_pkey_col . ';name',
+
+      'table_alias_mapping' => [$table_alias => $linker_table],
+      'as' => 'contact_name',
+    ]);
+
+    // The contact description
+    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'contact_description', $terms['description'], $max_lengths['description'], [
+      'action' => 'read_value',
+      'drupal_store' => FALSE,
+      'path' => $table_alias . '.' . $linker_fkey_column . '>' . $object_table . '.' . $object_pkey_col . ';description',
+      'table_alias_mapping' => [$table_alias => $linker_table],
+      'as' => 'contact_description',
+    ]);
+
+    // The type of contact
+    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'contact_type', $terms['contact_type'], $max_lengths['contact_type'], [
+      'action' => 'read_value',
+      'drupal_store' => FALSE,
+      'path' => $table_alias . '.' . $linker_fkey_column . '>' . $object_table . '.' . $object_pkey_col
+        . ';' . $object_table . '.type_id>cvterm.cvterm_id;name',
+      'table_alias_mapping' => [$table_alias => $linker_table],
+      'as' => 'contact_type',
```

Comment:
```suggestion
    $table_alias = $linker_table . '_' . preg_replace( '/[^a-z0-9]+/', '', strtolower( $term ) );
    $table_alias_mapping = [$table_alias => $linker_table];

    // FINALLY, THE PROPERTIES!
    $properties = [];

    // Define the base table record id.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'record_id', $terms['base_pkey'], [
      'action' => 'store_id',
      'drupal_store' => TRUE,
      'path' => $base_table . '.' . $base_pkey_col,
    ]);

    // This property will store the Drupal entity ID of the linked chado
    // record, if one exists.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'entity_id', self::$drupal_entity_term, [
      'action' => 'function',
      'drupal_store' => TRUE,
      'namespace' => self::$chadostorage_namespace,
      'function' => self::$drupal_entity_callback,
      'ftable' => self::$object_table,
      'fkey' => self::$object_id,
    ]);

    // Define the linker table that links the base table to the object table.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_id', $terms['linker_pkey'], [
      'action' => 'store_pkey',
      'drupal_store' => TRUE,
      'path' => $table_alias . '.' . $linker_pkey_col,
      'table_alias_mapping' => $table_alias_mapping,
    ]);

    // Define the link between the base table and the linker table.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'link', $terms['linker_left'], [
      'action' => 'store_link',
      'drupal_store' => TRUE,
      'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $linker_left_col,
      'table_alias_mapping' => $table_alias_mapping,
    ]);

    // Define the link between the linker table and the object table.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, $linker_fkey_column, $terms['linker_right'], [
      'action' => 'store',
      'drupal_store' => TRUE,
      'path' => $table_alias . '.' . $linker_fkey_column,
      'table_alias_mapping' => $table_alias_mapping,
      'delete_if_empty' => TRUE,
      'empty_value' => 0,
    ]);

    // Linker type_id
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_type_id', $terms['linker_type_id'], [
      'action' => 'store',
      'drupal_store' => FALSE,
      'path' => $table_alias . '.type_id',
      'table_alias_mapping' => $table_alias_mapping,
      'as' => 'linker_type_id',
    ]);

    // Linker Rank
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_rank', $terms['linker_rank'], [
      'action' => 'store',
      'drupal_store' => FALSE,
      'path' => $table_alias . '.rank',
      'table_alias_mapping' => $table_alias_mapping,
      'as' => 'linker_rank',
    ]);

    // The object table, the destination table of the linker table
    // The contact name
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'contact_name', $terms['name'], $max_lengths['name'], [
      'action' => 'read_value',
      'drupal_store' => FALSE,
      'path' => $table_alias . '.' . $linker_fkey_column . '>' . $object_table . '.' . $object_pkey_col . ';name',

      'table_alias_mapping' => $table_alias_mapping,
      'as' => 'contact_name',
    ]);

    // The contact description
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'contact_description', $terms['description'], $max_lengths['description'], [
      'action' => 'read_value',
      'drupal_store' => FALSE,
      'path' => $table_alias . '.' . $linker_fkey_column . '>' . $object_table . '.' . $object_pkey_col . ';description',
      'table_alias_mapping' => $table_alias_mapping,
      'as' => 'contact_description',
    ]);

    // The type of contact
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'contact_type', $terms['contact_type'], $max_lengths['contact_type'], [
      'action' => 'read_value',
      'drupal_store' => FALSE,
      'path' => $table_alias . '.' . $linker_fkey_column . '>' . $object_table . '.' . $object_pkey_col
        . ';' . $object_table . '.type_id>cvterm.cvterm_id;name',
      'table_alias_mapping' => $table_alias_mapping,
      'as' => 'contact_type',
```
What do you think about this change? I feel it simplifies reading the code not having this same array "hardcoded" several times

Reply:
I see :-) I'm actually about to make quite a big change here to inherit a lot of this from the original contact field to reduce code duplication and make it really clear what the difference between the two fields is :-) But I'll keep this change in mind while I do it!